### PR TITLE
[RFC] Remove "visual duplication" of ModelFactory::new()->create()

### PR DESCRIPTION
**Pull request by wouterj**
This is an approach on removing "visual duplication" when creating new models:

```php
// before
$post = PostFactory::new()->create();
$post = PostFactory::new()->published()->create();
$posts = PostFactory::new()->createMany(3);
$posts = PostFactory::new()->published()->createMany(3);

// after
$post = PostFactory::createOne();
$post = PostFactory::new()->published()->create();
$posts = PostFactory::createMany(3);
$posts = PostFactory::new()->published()->many(3)->create();
```

We can't use `PostFactory::create()` as it conflicts with the `Factory#create()` instance method. <s>I used the magic method name to clearly indicate that we're not creating a new factory, but we are creating a new entity. An alternative name that doesn't need the magic `__callStatic()` method can be `createSingle()` (which seems consistent with `createMany()`).</s>

### CODE REVIEW

**Comment by kbond** [comment]
I think I like `createSingle()` as it fully removes the "visual duplication".

Now that `createMany()` is just an alias for `->many()->create()` what about making it static so:

```php
PostFactory::createSingle(): Post
PostFactory::createMany(): Post[]
```

Could this be handled w/o a bc break with some `__call()/`__callStatic()` trickery?

**Comment by kbond** [comment]
Or what about `createOne()`? Would sort of match `findOneBy` in doctrine...

**Comment by wouterj** [comment]
I've just updated everything to `createSingle()`, but I'm +1 for `createOne()` :) 

> Now that createMany() is just an alias for ->many()->create() what about making it static so [...]

I've included this change in the PR and deprecated the `createMany()` instance method. Let me know what you think.

As you see, the build breaks because of the `create_many()` helper function. I'm not sure what you want to do here, deprecate the function? (I can't see how to implement this function without a `createMany()` instance method)

File: src/Bundle/Resources/skeleton/Factory.tpl.php [hunk 1]
```diff
@@ -10,6 +10,7 @@
 use Zenstruck\Foundry\Proxy;
 
 /**
+ * @method static <?= $entity->getShortName() ?>|Proxy createSingle(array $attributes = [])
```
Comment (kbond) [inline src/Bundle/Resources/skeleton/Factory.tpl.php:13]: `createMany` below should be marked as `static` now, right?

Comment (kbond) [reply src/Bundle/Resources/skeleton/Factory.tpl.php:13]: Or will that typehint interfere with the non-static createMany() `@method` on `Factory`?

**Comment by kbond** [comment]
> build breaks because of the create_many() helper function. I'm not sure what you want to do here

What about:

```php
function create_many(int $number, string $class, $attributes = []): array
{
    return factory($class)->many($number)->create($attributes);
}
```

File: src/Bundle/Resources/skeleton/Factory.tpl.php (hunk 1 above)
Comment (wouterj) [reply src/Bundle/Resources/skeleton/Factory.tpl.php]: I've now made it static, didn't try autocompletion yet (I don't normally use an IDE when working on small projects like this)

Comment (kbond) [reply src/Bundle/Resources/skeleton/Factory.tpl.php]: Regarding auto-completion, I think the'll be some issues mixing static and non-static `@method`'s of the same name. At least with PhpStorm.

Comment (wouterj) [reply src/Bundle/Resources/skeleton/Factory.tpl.php]: As the instance method is deprecated, should we only document the static method in the PHPdoc?

Comment (wouterj) [reply src/Bundle/Resources/skeleton/Factory.tpl.php]: I've updated the PR with my latest suggestion

File: src/ModelFactory.php [hunk 2]
```diff
@@ -6,6 +6,9 @@
  * @template TModel of object
  * @template-extends Factory<TModel>
  *
+ * @method static Proxy[]|object[] createMany(int $number, array|callable $attributes = []): array
```
Comment (kbond) [inline src/ModelFactory.php:9]: I think the trailing `: array` can be removed.

File: src/Bundle/Resources/skeleton/Factory.tpl.php (hunk 1 above)
Comment (kbond) [reply src/Bundle/Resources/skeleton/Factory.tpl.php]: Not ideal but I think this is best.

**Comment by kbond** [comment]
I love how the tests read now (after replacing `::new()->create()` with `::createOne()`)!

**Review by weaverryan — APPROVED** [review]
I love this.

no more new() call for most cases (where you’re just creating one or many and maybe passing in some attributes). Then, the longer ->new() only when you need extra config/setup for your factory.

**Comment by wouterj** [comment]
Updated the PR to also update the README with the new API. I didn't find any CHANGELOG file in the repository, I guess you're doing that manually while releasing?

The actions fail because of 2 uncovered lines of code (the bad method call exceptions). I don't think these should be tested, but if you want I can add a test.

**Comment by kbond** [comment]
Awesome, looks great!

> I didn't find any CHANGELOG file in the repository, I guess you're doing that manually while releasing?

Yeah, I use the Github releases as a changelog right now.

> The actions fail because of 2 uncovered lines of code

No worries - I only use these as a loose metric - I'm not strict about it. I've been meaning to increase the "passable" threshold from 0%

File: README.md [hunk 3]
```diff
@@ -286,22 +290,6 @@ $posts = PostFactory::randomSet(4); // array containing 4 "Post|Proxy" objects
 $posts = PostFactory::randomRange(0, 5); // array containing 0-5 "Post|Proxy" objects
 ```
 
-### Instantiate your `ModelFactory`
```
Comment (Nyholm) [inline README.md:289]: Can I run `new PostFactory()` now? Or why was this removed?

**Review by Nyholm — APPROVED** [review]

File: README.md (hunk 3 above)
Comment (wouterj) [reply README.md:289]: No, you still have to use `::new()`. However, you'll only use `new()` in complex cases, otherwise you use `Factory::createOne(...)` or `Factory::createMany(...)`. So I removed the main section about this and instead moved this into the complex code examples (line 331 and 361)

**Comment by kbond** [comment]
Thanks @wouterj!

**Review by Nyholm — COMMENTED** [review]
(submitting my comment I wrote yestetday)
